src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,218 @@
+
+    HINLINE void operator()( FieldTmp* fieldTmp,
+                             const uint32_t currentStep) const
+    {
+        DataConnector &dc = Environment<>::get().DataConnector();
+
+        /* load species without copying the particle data to the host */
+        SpeciesName* speciesTmp = &(dc.getData<SpeciesName >(SpeciesName::FrameType::getName(), true));
+
+        /* run algorithm */
+        typedef typename CreateDensityOperation<SpeciesName>::type::Solver ChargeDensitySolver;
+        fieldTmp->computeValue < area, ChargeDensitySolver > (*speciesTmp, currentStep);
+        dc.releaseData(SpeciesName::FrameType::getName());
+    }
+};
+
+} // namespace detail
+
+void ChargeConservation::notify(uint32_t currentStep)
+{
+    typedef SuperCellSize BlockDim;
+
+    DataConnector &dc = Environment<>::get().DataConnector();
+
+    /* load FieldTmp without copy data to host */
+    FieldTmp* fieldTmp = &(dc.getData<FieldTmp > (FieldTmp::getName(), true));
+    /* reset density values to zero */
+    fieldTmp->getGridBuffer().getDeviceBuffer().setValue(FieldTmp::ValueType(0.0));
+
+    /* calculate and add the charge density values from all species in FieldTmp */
+    ForEach<VectorAllSpecies, picongpu::detail::ComputeChargeDensity<bmpl::_1,bmpl::int_<CORE + BORDER> >, MakeIdentifier<bmpl::_1> > computeChargeDensity;
+    computeChargeDensity(forward(fieldTmp), currentStep);
+
+    /* add results of all species that are still in GUARD to next GPUs BORDER */
+    EventTask fieldTmpEvent = fieldTmp->asyncCommunication(__getTransactionEvent());
+    __setTransactionEvent(fieldTmpEvent);
+
+    /* cast libPMacc Buffer to cuSTL Buffer */
+    BOOST_AUTO(fieldTmp_coreBorder,
+                 fieldTmp->getGridBuffer().
+                 getDeviceBuffer().cartBuffer().
+                 view(BlockDim::toRT(), -BlockDim::toRT()));
+
+    /* cast libPMacc Buffer to cuSTL Buffer */
+    BOOST_AUTO(fieldE_coreBorder,
+                 dc.getData<FieldE > (FieldE::getName(), true).getGridBuffer().
+                 getDeviceBuffer().cartBuffer().
+                 view(BlockDim::toRT(), -BlockDim::toRT()));
+
+    /* run calculation: fieldTmp = | div E * eps_0 - rho | */
+    using namespace lambda;
+    using namespace PMacc::math::math_functor;
+    typedef picongpu::detail::Div<simDim, typename FieldTmp::ValueType> myDiv;
+    algorithm::kernel::Foreach<BlockDim>()
+        (fieldTmp_coreBorder.zone(), fieldTmp_coreBorder.origin(),
+        cursor::make_NestedCursor(fieldE_coreBorder.origin()),
+            _1 = _abs(expr(myDiv())(_2) * EPS0 - _1));
+
+    /* reduce charge derivation (fieldTmp) to get the maximum value */
+    typename FieldTmp::ValueType maxChargeDiff =
+        algorithm::kernel::Reduce()
+            (fieldTmp_coreBorder.origin(), fieldTmp_coreBorder.zone(), PMacc::nvidia::functors::Max());
+
+    /* reduce again across mpi cluster */
+    container::HostBuffer<typename FieldTmp::ValueType, 1> maxChargeDiff_host(1);
```

Comment:
You need a `__getTransactionEvent().waitForFisnished();` before this line, because CuSTL assign is set no event system hints and is not race condition save.